lib/active_shipping/carriers/correios.rb

```diff
@@ -0,0 +1,271 @@
+      81868 => '(Grupo 1) e-SEDEX, com contrato',
+      81833 => '(Grupo 2) e-SEDEX, com contrato',
+      81850 => '(Grupo 3) e-SEDEX, com contrato'
+    }.freeze
+
+    protected
+
+    def perform(urls)
+      urls.map { |url| ssl_get(url) }
+    end
+
+    class CorreiosRequest
+
+      URL = "http://ws.correios.com.br/calculador/CalcPrecoPrazo.aspx"
+
+      RETURN_TYPE = 'xml'
+      RETURN_INFORMATION_TYPE = {
+        :prices => '1',
+        :time => '2',
+        :prices_and_time => '3'
+      }
+
+      attr_reader :origin, :destination, :urls
+
+      def initialize(origin, destination, packages, options)
+        @options = options
+        @origin = origin
+        @destination = destination
+
+        packages = packages.map do |package|
+          CorreiosPackage.new(package, 1)
+        end
+
+        @params = {
+          company_id: options[:company_id],
+          password: options[:password],
+          service_type: service_type,
+          origin_zip: origin.zip,
+          destination_zip: destination.zip,
+          mao_propria_extra: parse_boolean(options[:mao_propria_extra]),
+          declared_value_extra: parse_currency(options[:declared_value_extra]),
+          delivery_notice_extra: parse_boolean(options[:delivery_notice_extra]),
+          return_type: RETURN_TYPE,
+          return_information: RETURN_INFORMATION_TYPE[:prices]
+        }
+        @urls = packages.map { |package| create_url(package) }
+      end
+
+      def create_response(raw_xmls)
+        correios_response = CorreiosResponse.new(self, raw_xmls)
+        correios_response.rate_response
+      end
+
+      private
+
+      def parse_boolean(param)
+        param == true ? 'S' : 'N'
+      end
+
+      def parse_currency(param)
+        param.nil? ? '0' : param.to_s.gsub('.', ',')
+      end
+
+      def params(package)
+        @params.merge(package.params)
+      end
+
+      def query_string(params)
+        "nCdEmpresa=#{params[:company_id]}&" +
+        "sDsSenha=#{params[:password]}&" +
+        "nCdServico=#{params[:service_type]}&" +
+        "sCepOrigem=#{params[:origin_zip]}&" +
+        "sCepDestino=#{params[:destination_zip]}&" +
+        "nVlPeso=#{params[:weight]}&" +
+        "nCdFormato=#{params[:format]}&" +
+        "nVlComprimento=#{params[:length]}&" +
+        "nVlAltura=#{params[:height]}&" +
+        "nVlLargura=#{params[:width]}&" +
+        "nVlDiametro=#{params[:diameter]}&" +
+        "sCdMaoPropria=#{params[:mao_propria_extra]}&" +
+        "nVlValorDeclarado=#{params[:declared_value_extra]}&" +
+        "sCdAvisoRecebimento=#{params[:delivery_notice_extra]}&" +
+        "nIndicaCalculo=#{params[:return_information]}&" +
+        "StrRetorno=#{params[:return_type]}"
+      end
+
+      def create_url(package)
+        "#{URL}?#{query_string(params(package))}"
+      end
+
+      def service_type
+        @options[:services].nil? ? '41106,40010' : @options[:services].join(',')
```

Comment:
Maybe use a constant for this default value, so it's a bit more clear what it means?